base/random.jl

```diff
@@ -1213,14 +1213,14 @@ let Floats = Union{Float16,Float32,Float64}
             $randfun!(A::AbstractArray) = $randfun!(GLOBAL_RNG, A)
 
             # generating arrays
-            $randfun{T}(rng::AbstractRNG, ::Type{T}, dims::Dims)       = $randfun!(rng, Array{T}(dims))
-            $randfun{T}(rng::AbstractRNG, ::Type{T}, dims::Integer...) = $randfun!(rng, Array{T}(dims...))
-            $randfun{T}(                  ::Type{T}, dims::Dims)       = $randfun(GLOBAL_RNG, T, dims)
-            $randfun{T}(                  ::Type{T}, dims::Integer...) = $randfun(GLOBAL_RNG, T, dims...)
-            $randfun(   rng::AbstractRNG,            dims::Dims)       = $randfun(rng, Float64, dims)
-            $randfun(   rng::AbstractRNG,            dims::Integer...) = $randfun(rng, Float64, dims...)
-            $randfun(                                dims::Dims)       = $randfun(GLOBAL_RNG, Float64, dims)
-            $randfun(                                dims::Integer...) = $randfun(GLOBAL_RNG, Float64, dims...)
+            $randfun{T}(rng::AbstractRNG, ::Type{T}, dims::Dims                     ) = $randfun!(rng, Array{T}(dims))
+            $randfun{T}(rng::AbstractRNG, ::Type{T}, dim0::Integer, dims::Integer...) = $randfun!(rng, Array{T}(dim0, dims...))
+            $randfun{T}(                  ::Type{T}, dims::Dims                     ) = $randfun(GLOBAL_RNG, T, dims)
+            $randfun{T}(                  ::Type{T}, dim0::Integer, dims::Integer...) = $randfun(GLOBAL_RNG, T, dim0, dims...)
```

Comment:
Ah, I missed the definition on line 1203.  I would prefer to just delete that definition.  (If there is some performance reason to define the one-argument case separately ... splatting penalty? ... we should still define it generically for all types, not just `T<:$Floats`.)

My reason is that, to add `randn` for a new type `T`, you should just have to define one method: `randn(rng::AbstractRNG, ::Type{T})`.   You shouldn't have to define other methods that we could get just by loosening the type signature.

If the price for that is slightly more obscure method errors, that's fine.   In fact, `no method matching randn(::MersenneTwister, ::Type{Int64})` is actually *more* helpful, because it tells you exactly what method you need to implement to make everything work.